### PR TITLE
Updates and fixes to history zeroing and flushing

### LINKED ISSUE #1207: FATES_PATCHAREA_AP only has data in chronologically-first history file
As of `sci.1.73.0_api.35.0.0`, the per-age-class patch area output `FATES_PATCHAREA_AP` only has data saved in the first (chronologically) history output file. The variable itself is there, but full of missing values.

This didn't happen in `sci.1.71.0_api.33.0.0`. I can't test any intervening tags because those are the closest ones with which we have CTSM working.

This blocks my ability to work on #1205.

### PR DESCRIPTION
**Pull request by rgknox**
### Description:

Some multi-dimensioned history variables were not being reported correctly because they were not being zero'd.  These variables were likely incremented over a patch-cohort loop, but being incremented on top of an uninitialized number.  This set of changes seeks to make the flushing and zeroing more consistent across all of the different groups of history variables. 

This also reduces confusion by moving flushing and zeroing to occur inside the routine that performs the updates.  These flushing and zeroing calls still happen redundantly in the host models, should be removed, but are not a liability to generating incorrect results. These calls are just superfluous.

Fixes #1207

### Collaborators:
@samsrabin 

### Checklist

*If this is your first time contributing, please read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/main/CONTRIBUTING.md) document.*

All checklist items must be checked to enable merging this pull request:

*Contributor*
- [ ] The in-code documentation has been updated with descriptive comments
- [ ] The documentation has been assessed to determine if updates are necessary

*Integrator*
- [ ] FATES PASS/FAIL regression tests were run
- [ ] Evaluation of test results for answer changes was performed and results provided

### Documentation

- [Technical Note](https://github.com/NGEET/fates-docs) update:
- [User's Guide](https://github.com/NGEET/fates-users-guide) update: 

### Test Results:

*CTSM (or) E3SM (specify which) test hash-tag:*

*CTSM (or) E3SM (specify which) baseline hash-tag:*

*FATES baseline hash-tag:*

*Test Output:*

### CODE REVIEW

**Comment by samsrabin** [comment]
Is there an addition or change we can/should make to the CTSM test list to avoid a bug like this in the future? E.g., I first noticed this with `FATES_PATCHAREA_AP`—is it not on any output file in the tests?

**Comment by rgknox** [comment]
There are other things to do with cleaning up the history files, but they are out of scope, so I think this PR is feature complete and ready for review. 
 I ran a one-off test and it passsed ERS_Ld10.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdAllVars

**Comment by rgknox** [comment]
@samsrabin we periodically need to update the FatesColdAllVars test to make sure it has all of the variables.  We have this script: [tools/FindInactive.py](https://github.com/NGEET/fates/blob/main/tools/FindInactive.py)  that trawls the code and updates the list, seems like we need to update it.

**Review by glemieux — APPROVED** [review]
One note and clarification on possible cleanup.  I'll start testing on this as soon as derecho is back up.

**Comment by glemieux** [comment]
regression testing underway

**Comment by glemieux** [comment]
Regression testing on cheyenne shows that most of the tests are returning B4B or the only diffs are in the fill values.  That said, the following testmods are return DIFFS:
```
FAIL ERP_Ld9.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdAllVars BASELINE fates-sci.1.76.3_api.35.1.0-ctsm5.2.007: DIFF
FAIL ERS_D_Ld30.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdPRT2 BASELINE fates-sci.1.76.3_api.35.1.0-ctsm5.2.007: DIFF
FAIL ERS_D_Ld5.1x1_brazil.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdHydro BASELINE fates-sci.1.76.3_api.35.1.0-ctsm5.2.007: DIFF
FAIL ERS_Lm12.1x1_brazil.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesFireLightningPopDens BASELINE fates-sci.1.76.3_api.35.1.0-ctsm5.2.007: DIFF
FAIL SMS_Lm1.f45_f45_mg37.I2000Clm60FatesSpCruRsGs.derecho_intel.clm-FatesColdBasic BASELINE fates-sci.1.76.3_api.35.1.0-ctsm5.2.007: DIFF
FAIL SMS_Lm3_D_Mmpi-serial.1x1_brazil.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdHydro BASELINE fates-sci.1.76.3_api.35.1.0-ctsm5.2.007: DIFF
```

I need to take a look at these to see if they make sense, but there two that jump out at me at first glance.  The `ERP_Ld9.f45_f45_mg37.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesColdAllVars` and `ERS_Lm12.1x1_brazil.I2000Clm50FatesCruRsGs.derecho_intel.clm-FatesFireLightningPopDens` tests returns huge DIFFS in `FATES_MORTALITY_PF` due to the new values being huge (i.e. `1e+38`).  Spitfire is set to mode `1` in the `AllVars` test and tree damage is on as well.  Spitfire is set to mode `4`.

File: main/FatesHistoryInterfaceMod.F90 [hunk 1]
```diff
@@ -5187,13 +5222,16 @@ subroutine update_history_hifrq2(this,nc,nsites,sites,bc_in,bc_out,dt_tstep)
...
-      call this%flush_hvars(nc,upfreq_in=group_hifr_complx)
 
+      ! Move this to the interface for consistency (rgk 0524)
+      call this%flush_hvars(nc,upfreq_in=group_hifr_complx)
+      
       dt_tstep_inv = 1.0_r8/dt_tstep
 
       do_sites: do s = 1,nsites
 
+         call this%zero_site_hvars(sites(s), upfreq_in=group_hifr_complx)
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesHistoryInterfaceMod.F90:5233]: good catch, yeah, there is a duplicate there, Ill remove

**Comment by rgknox** [comment]
The dynamics time-scale complex variables were not being flushed and zero'd correctly, this was corrected, the MORTALITY_PF variable is an example of that

**Comment by glemieux** [comment]
> The dynamics time-scale complex variables were not being flushed and zero'd correctly, this was corrected, the MORTALITY_PF variable is an example of that

Per our conversation, the diff that I'm seeing may be doing something on top of the fill value?
```
FATES_MORTALITY_PF   (lon,lat,fates_levpft,time)  t_index =      2     2
       12996    39744  (    33,     2,     1,     1) (    23,    21,    10,     1) (    33,     2,     1,     1) (    33,     2,     1,     1)
                12996   1.225934268493151E-01          2.799785209598797E-03 1.1E+38  1.225934268493151E-01 1.0E+00  1.225934268493151E-01
                12996   1.050000000000002E+38          1.050000000000002E+38          1.050000000000002E+38          1.050000000000002E+38
                39744  (    33,     2,     1,     1) (    33,     2,     1,     1)
           avg abs field values:    6.526914797498996E-02    rms diff: 1.0E+38   avg rel diff(npos):  1.0E+00
                                    1.050000000000126E+38                        avg decimal digits(ndif):  0.0 worst:  0.0
RMS FATES_MORTALITY_PF               1.0500E+38            NORMALIZED  2.0000E+00
```

The test location on derecho: `/glade/u/home/glemieux/scratch/ctsm-tests/tests_pr1208-fates`

**Comment by glemieux** [comment]
Per our conversation, I went back and looked and realized I misread the `cprnc` output.  I keep thinking the second line below is the current commit; it's actually the baseline:

```
FATES_MORTALITY_PF   (lon,lat,fates_levpft,time)  t_index =      2     2
       12996    39744  (    33,     2,     1,     1) (    23,    21,    10,     1) (    33,     2,     1,     1) (    33,     2,     1,     1)
                12996   1.225934268493151E-01          2.799785209598797E-03 1.1E+38  1.225934268493151E-01 1.0E+00  1.225934268493151E-01
                12996   1.050000000000002E+38          1.050000000000002E+38          1.050000000000002E+38          1.050000000000002E+38
                39744  (    33,     2,     1,     1) (    33,     2,     1,     1)
           avg abs field values:    6.526914797498996E-02    rms diff: 1.0E+38   avg rel diff(npos):  1.0E+00
                                    1.050000000000126E+38                        avg decimal digits(ndif):  0.0 worst:  0.0
RMS FATES_MORTALITY_PF               1.0500E+38            NORMALIZED  2.0000E+00
```

So this is indeed corrected as you explained.  I think we can integrate.

For postertity, the test location on derecho: `/glade/u/home/glemieux/scratch/ctsm-tests/tests_pr1208-fates`

**Pull request merged**